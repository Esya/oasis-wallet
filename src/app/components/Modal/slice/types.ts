export interface Modal {
  title: string
  description: string
  handleConfirm: () => void

  /**
   * Is this a dangerous operation?
   *
   * If marked as such, it will only be possible to execute it if the wallet is configured to run in dangerous mode.
   *
   * It also automatically implies a mandatory waiting time of 10 sec, unless specified otherwise.
   */
  isDangerous: boolean

  /**
   * How long does the user have to wait before he can actually confirm the action?
   */
  mustWaitSecs?: number
}

export interface ModalState {
  current: Modal | null
  stash: Modal[]
}
