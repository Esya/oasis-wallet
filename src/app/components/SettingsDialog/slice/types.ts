export interface SettingsState {
  allowDangerous: boolean
}
