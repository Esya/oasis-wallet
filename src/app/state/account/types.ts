import { Transaction } from 'app/state/transaction/types'

export interface BalanceDetails {
  total: number
  debonding: number
  available: number
  delegations: number
}

export interface Account {
  address: string
  liquid_balance: number
}

/* --- STATE --- */
export interface AccountState extends Account {
  loading: boolean
  accountError: string | null
  transactions: Transaction[]
  transactionsError: string | null
}
