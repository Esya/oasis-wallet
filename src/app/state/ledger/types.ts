import { ErrorPayload } from 'types/errors'
import { WalletBalance } from '../wallet/types'

/* --- STATE --- */
export interface LedgerAccount {
  publicKey: string
  address: string
  path: number[]
  balance: WalletBalance
  selected: boolean
}

export enum LedgerStep {
  OpeningUSB = 'opening_usb',
  LoadingAccounts = 'loading_accounts',
  LoaddingBalances = 'loading_balances',
  Done = 'done',
}

export interface LedgerState {
  accounts: LedgerAccount[]
  step?: LedgerStep
  error?: ErrorPayload
}
