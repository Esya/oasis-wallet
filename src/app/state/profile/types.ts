/* --- STATE --- */
export interface ProfileState {
  profileModalOpen: boolean
}
