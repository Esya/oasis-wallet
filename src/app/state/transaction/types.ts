import { ErrorPayload } from 'types/errors'

export enum TransactionType {
  StakingTransfer = 'staking.Transfer',
  StakingAddEscrow = 'staking.AddEscrow',
  StakingReclaimEscrow = 'staking.ReclaimEscrow',
  StakingAmendCommissionSchedule = 'staking.AmendCommissionSchedule',
  StakingAllow = 'staking.Allow',
  StakingWithdraw = 'staking.Withdraw',
  RoothashExecutorCommit = 'roothash.ExecutorCommit',
  RoothashExecutorProposerTimeout = 'roothash.ExecutorProposerTimeout',
  RegistryRegisterEntity = 'registry.RegisterEntity',
  RegistryRegisterNode = 'registry.RegisterNode',
  RegistryRegisterRuntime = 'registry.RegisterRuntime',
  GovernanceCastVote = 'governance.CastVote',
  GovernanceSubmitProposal = 'governance.SubmitProposal',
  BeaconPvssCommit = 'beacon.PVSSCommit',
  BeaconPvssReveal = 'beacon.PVSSReveal',
  BeaconVrfProve = 'beacon.VRFProve',
  ConsensusDeposit = 'consensus.Deposit',
  ConsensusWithdraw = 'consensus.Withdraw',
  ConsensusAccountsParameters = 'consensus.Parameters',
  ConsensusBalance = 'consensus.Balance',
  ConsensusAccount = 'consensus.Account',
}

export interface Transaction {
  amount: number | undefined
  fee: number | undefined
  from: string | undefined
  hash: string
  level: number | undefined
  status: boolean | undefined
  timestamp: number | undefined
  to: string | undefined
  type: TransactionType
  // These are undefined on consensus transaction
  // Only appear on ParaTime transaction
  runtimeName: string | undefined
  runtimeId: string | undefined
  round: number | undefined
}

/* --- STATE --- */
export interface TransactionState {
  step?: TransactionStep
  active: boolean
  success: boolean
  error?: ErrorPayload
  preview?: TransactionPreview
}

export type NewTransactionType = 'transfer' | 'addEscrow' | 'reclaimEscrow'
export type TransactionPayload = TransferPayload | AddEscrowPayload | ReclaimEscrowPayload

export interface TransactionPreview {
  transaction: TransactionPayload
  fee?: string
  gas?: string
}

/**
 * Flow for a transaction to be sent.
 */
export enum TransactionStep {
  Building = 'building',
  Preview = 'preview',
  Signing = 'signing',
  Submitting = 'submitting',
  Sent = 'sent',
}

export interface TransactionSent {
  from: string
  to: string
  amount: number
}

export interface TransferPayload {
  type: 'transfer'

  /* bech32 Address */
  to: string

  /* Token amount */
  amount: number
}

export interface AddEscrowPayload {
  type: 'addEscrow'

  /* bech32 Address */
  validator: string

  /* Token amount */
  amount: number
}

export interface ReclaimEscrowPayload {
  type: 'reclaimEscrow'

  /* bech32 Address */
  validator: string

  /* Shares to be reclaimed */
  shares: number

  /* Amount to be reclaimed */
  amount: number
}
