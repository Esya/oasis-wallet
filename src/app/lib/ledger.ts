import { ContextSigner } from '@oasisprotocol/client/dist/signature'
import OasisApp from '@oasisprotocol/ledger'
import { Wallet, WalletType } from 'app/state/wallet/types'
import { WalletError, WalletErrors } from 'types/errors'
import { hex2uint } from './helpers'

interface Response {
  return_code: number
  error_message: string
  [index: string]: unknown
}

interface LedgerAccount {
  publicKey: Uint8Array
  path: number[]
}

const successOrThrow = (response: Response, message: string) => {
  if (response.return_code !== 0x9000) {
    switch (response.return_code) {
      case 25600:
        throw new WalletError(WalletErrors.LedgerAppVersionNotSupported, response.error_message)
      case 27014:
        throw new WalletError(WalletErrors.LedgerTransactionRejected, response.error_message)
      case 26628:
        throw new WalletError(WalletErrors.LedgerCannotOpenOasisApp, response.error_message)

      default:
        throw new WalletError(WalletErrors.LedgerUnknownError, response.error_message)
    }
  }
  return response
}
export class Ledger {
  public static async enumerateAccounts(transport: any, count = 5) {
    const accounts: LedgerAccount[] = []

    try {
      const app = new OasisApp(transport)
      for (let i = 0; i < count; i++) {
        const path = [44, 474, 0, 0, i]
        const publicKeyResponse = successOrThrow(await app.publicKey(path), 'ledger public key')
        accounts.push({ path, publicKey: new Uint8Array(publicKeyResponse.pk as Buffer) })
      }
    } catch (e) {
      throw e
    }

    return accounts
  }
}

export class LedgerSigner implements ContextSigner {
  // Oasis ledger app. Some typings would be great...!
  protected transport?: any
  protected path: number[]
  protected publicKey: Uint8Array

  constructor(wallet: Wallet) {
    if (!wallet.path || wallet.type !== WalletType.Ledger) {
      throw new Error('Given wallet is not a ledger wallet')
    }
    this.path = wallet.path
    this.publicKey = hex2uint(wallet.publicKey)
  }

  public setTransport(transport: any) {
    this.transport = transport
  }

  public public(): Uint8Array {
    return this.publicKey
  }

  public async sign(context: string, message: Uint8Array): Promise<Uint8Array> {
    if (!this.transport) {
      throw new Error('Cannot sign using ledger without an active WebUSB transport')
    }

    const app = new OasisApp(this.transport)
    const response = successOrThrow(await app.sign(this.path, context, Buffer.from(message)), 'ledger sign')
    return new Uint8Array(response.signature as Buffer)
  }
}
