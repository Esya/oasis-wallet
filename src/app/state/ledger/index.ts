import { PayloadAction } from '@reduxjs/toolkit'
import { ErrorPayload } from 'types/errors'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { useInjectReducer, useInjectSaga } from 'utils/redux-injectors'
import { ledgerSaga } from './saga'
import { LedgerAccount, LedgerState, LedgerStep } from './types'

export const initialState: LedgerState = { accounts: [] }

const slice = createSlice({
  name: 'ledger',
  initialState,
  reducers: {
    clear(state, action: PayloadAction<void>) {
      state.accounts = []
      state.derivationPathType = undefined
      state.error = undefined
    },
    enumerateAccounts(state, action: PayloadAction<string>) {
      state.step = undefined
      state.derivationPathType = action.payload
      state.accounts = []
    },
    toggleAccount(state, action: PayloadAction<number>) {
      const index = action.payload
      state.accounts[index].selected = !state.accounts[index].selected
    },
    accountsListed(state, action: PayloadAction<LedgerAccount[]>) {
      state.accounts = action.payload
    },
    setStep(state, action: PayloadAction<LedgerStep>) {
      state.step = action.payload
    },
    setDerivationPathType(state, action: PayloadAction<string>) {
      state.derivationPathType = action.payload
      state.accounts = []
    },
    operationFailed(state, action: PayloadAction<ErrorPayload>) {
      state.error = action.payload
      state.step = undefined
    },
  },
})

export const { actions: ledgerActions } = slice

export const useLedgerSlice = () => {
  useInjectReducer({ key: slice.name, reducer: slice.reducer })
  useInjectSaga({ key: slice.name, saga: ledgerSaga })
  return { actions: slice.actions }
}
