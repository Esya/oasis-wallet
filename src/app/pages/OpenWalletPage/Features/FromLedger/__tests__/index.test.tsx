import { cleanup, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { ledgerActions } from 'app/state/ledger'
import { LedgerStep } from 'app/state/ledger/types'
import { walletActions } from 'app/state/wallet'
import * as React from 'react'
import { Provider, useDispatch } from 'react-redux'
import { configureAppStore } from 'store/configureStore'
import { ThemeProvider } from 'styles/theme/ThemeProvider'

import { FromLedger } from '..'

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useTranslation: () => {
    return {
      t: str => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    }
  },
  useDispatch: jest.fn(),
}))

const renderComponent = (store, abortFunction = () => {}) =>
  render(
    <Provider store={store}>
      <ThemeProvider>
        <FromLedger abort={abortFunction} />
      </ThemeProvider>
    </Provider>,
  )

describe('<FromLedger  />', () => {
  let store: ReturnType<typeof configureAppStore>
  let mockDispatch = useDispatch as jest.Mock<typeof useDispatch>

  beforeEach(() => {
    store = configureAppStore()
    mockDispatch.mockImplementation(() => jest.fn())
  })

  afterEach(() => {
    jest.resetModules()
    jest.resetAllMocks()
    cleanup()
    document.body.innerHTML = ''
  })

  it('should match snapshot', () => {
    const component = renderComponent(store)
    expect(component).toMatchSnapshot()
    screen.getByTestId('ledger-open-accounts')
  })

  it('should list the accounts when done', () => {
    const component = renderComponent(store)
    store.dispatch(
      ledgerActions.accountsListed([
        {
          address: 'oasis1qzyqaxestzlum26e2vdgvkerm6d9qgdp7gh2pxqe',
          balance: { available: '0', debonding: '0', escrow: '0', total: '0' },
          path: [44, 474, 0],
          publicKey: '00',
          selected: false,
        },
      ]),
    )

    store.dispatch(ledgerActions.setStep(LedgerStep.Done))
    expect(component.getByText('oasis1qzyq...7gh2pxqe')).toBeInTheDocument()
  })

  it('should open the selected accounts', () => {
    const dispatchFn = jest.fn()
    mockDispatch.mockImplementation(() => dispatchFn)

    renderComponent(store)
    store.dispatch(
      ledgerActions.accountsListed([
        {
          address: 'oasis1qzyqaxestzlum26e2vdgvkerm6d9qgdp7gh2pxqe',
          balance: { available: '0', debonding: '0', escrow: '0', total: '0' },
          path: [44, 474, 0],
          publicKey: '00',
          selected: false,
        },
        {
          address: 'oasis1qqv25adrld8jjquzxzg769689lgf9jxvwgjs8tha',
          balance: { available: '0', debonding: '0', escrow: '0', total: '0' },
          path: [44, 474, 1],
          publicKey: '00',
          selected: false,
        },
      ]),
    )

    store.dispatch(ledgerActions.setStep(LedgerStep.Done))
    userEvent.click(screen.getByText('oasis1qzyq...7gh2pxqe'))
    expect(dispatchFn).toHaveBeenLastCalledWith({ payload: 0, type: ledgerActions.toggleAccount.type })
    store.dispatch(ledgerActions.toggleAccount(0))

    userEvent.click(screen.getByTestId('ledger-open-accounts'))
    expect(dispatchFn).toHaveBeenLastCalledWith(
      expect.objectContaining({ type: walletActions.openWalletsFromLedger.type }),
    )
  })
})
