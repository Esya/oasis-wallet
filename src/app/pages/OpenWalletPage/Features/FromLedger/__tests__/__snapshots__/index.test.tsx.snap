// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FromLedgerModal  /> should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c18 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #0092f6;
  stroke: #0092f6;
}

.c18 g {
  fill: inherit;
  stroke: inherit;
}

.c18 *:not([stroke])[fill="none"] {
  stroke-width: 0;
}

.c18 *[stroke*="#"],
.c18 *[STROKE*="#"] {
  stroke: inherit;
  fill: none;
}

.c18 *[fill-rule],
.c18 *[FILL-RULE],
.c18 *[fill*="#"],
.c18 *[FILL*="#"] {
  fill: inherit;
  stroke: none;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  background-color: #FFFFFF;
  color: #444444;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 750px;
  padding: 24px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 100%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-right: 5px;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 440px;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-left: 12px;
  margin-right: 12px;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c21 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding-top: 48px;
}

.c23 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 12px;
}

.c19 {
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  font-size: 14px;
  line-height: 20px;
}

.c10 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  color: inherit;
  outline: none;
  border: none;
  padding: 0;
  text-align: inherit;
  opacity: 0.3;
  cursor: default;
}

.c10:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c10:focus > circle,
.c10:focus > ellipse,
.c10:focus > line,
.c10:focus > path,
.c10:focus > polygon,
.c10:focus > polyline,
.c10:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c10:focus::-moz-focus-inner {
  border: 0;
}

.c10:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c10:focus:not(:focus-visible) > circle,
.c10:focus:not(:focus-visible) > ellipse,
.c10:focus:not(:focus-visible) > line,
.c10:focus:not(:focus-visible) > path,
.c10:focus:not(:focus-visible) > polygon,
.c10:focus:not(:focus-visible) > polyline,
.c10:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c10:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c22 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #0092f6;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  opacity: 0.3;
  cursor: default;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  font-weight: bold;
}

.c22:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c22:focus > circle,
.c22:focus > ellipse,
.c22:focus > line,
.c22:focus > path,
.c22:focus > polygon,
.c22:focus > polyline,
.c22:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c22:focus::-moz-focus-inner {
  border: 0;
}

.c22:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c22:focus:not(:focus-visible) > circle,
.c22:focus:not(:focus-visible) > ellipse,
.c22:focus:not(:focus-visible) > line,
.c22:focus:not(:focus-visible) > path,
.c22:focus:not(:focus-visible) > polygon,
.c22:focus:not(:focus-visible) > polyline,
.c22:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c22:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c24 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  border: 2px solid #0092f6;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #0092f6;
  color: #f8f8f8;
  border-radius: 18px;
  opacity: 0.3;
  cursor: default;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  font-weight: bold;
}

.c24:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c24:focus > circle,
.c24:focus > ellipse,
.c24:focus > line,
.c24:focus > path,
.c24:focus > polygon,
.c24:focus > polyline,
.c24:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c24:focus::-moz-focus-inner {
  border: 0;
}

.c24:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c24:focus:not(:focus-visible) > circle,
.c24:focus:not(:focus-visible) > ellipse,
.c24:focus:not(:focus-visible) > line,
.c24:focus:not(:focus-visible) > path,
.c24:focus:not(:focus-visible) > polygon,
.c24:focus:not(:focus-visible) > polyline,
.c24:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c24:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c6 {
  margin-top: 0px;
  margin-bottom: 24px;
  font-size: 1;
  line-height: normal;
  max-width: 1200px;
  font-weight: 600;
}

.c20 {
  box-sizing: border-box;
  font-size: inherit;
  line-height: inherit;
  color: #0092f6;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c20:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c15 {
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: none;
  -webkit-appearance: none;
  background: transparent;
  color: inherit;
  width: 100%;
  padding: 11px;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0,0,0,0.33);
  border-radius: 4px;
  outline: none;
  border: none;
}

.c15::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c15::-moz-placeholder {
  color: #AAAAAA;
}

.c15:-ms-input-placeholder {
  color: #AAAAAA;
}

.c15::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c15::-moz-focus-inner {
  border: none;
  outline: none;
}

.c15:-moz-placeholder,
.c15::-moz-placeholder {
  opacity: 1;
}

.c14 {
  position: relative;
  width: 100%;
}

.c0 {
  font-family: Rubik;
  font-size: 18px;
  line-height: 24px;
  background-color: #ffffff;
  color: #444444;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  font-family: Rubik;
  font-size: 18px;
  line-height: 24px;
  background-color: #ffffff;
  color: #444444;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  background: transparent;
  position: relative;
  z-index: 20;
  pointer-events: none;
  outline: none;
  position: fixed;
  top: 0px;
  left: 0px;
  right: 0px;
  bottom: 0px;
}

.c2 {
  position: absolute;
  top: 0px;
  left: 0px;
  right: 0px;
  bottom: 0px;
  background-color: rgba(0,0,0,0.5);
  color: #f8f8f8;
  pointer-events: all;
  will-change: transform;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-height: 48px;
  background-color: #FFFFFF;
  color: #444444;
  outline: none;
  pointer-events: all;
  z-index: 20;
  position: absolute;
  max-height: calc(100% - 0px - 0px);
  max-width: calc(100% - 0px - 0px);
  border-radius: 4px;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  -webkit-animation: hWAzxx 0.2s ease-in-out forwards;
  animation: hWAzxx 0.2s ease-in-out forwards;
}

.c4 {
  width: 0;
  height: 0;
  overflow: hidden;
  position: absolute;
}

.c16 {
  cursor: default;
}

.c11 {
  border: 1px solid rgba(0,0,0,0.33);
  border-radius: 4px;
}

@media only screen and (max-width:768px) {
  .c5 {
    padding: 12px;
  }
}

@media only screen and (max-width:768px) {
  .c8 {
    padding-right: 5px;
  }
}

@media only screen and (max-width:768px) {
  .c17 {
    margin-left: 6px;
    margin-right: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c21 {
    padding-top: 24px;
  }
}

@media only screen and (max-width:768px) {
  .c23 {
    width: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c6 {
    margin-top: 0px;
  }
}

@media only screen and (max-width:768px) {
  .c6 {
    margin-bottom: 12px;
  }
}

@media only screen and (max-width:768px) {
  .c1 {
    position: fixed;
    width: 100%;
    height: 100%;
    min-height: 100vh;
  }
}

@media only screen and (max-width:768px) {
  .c2 {
    position: relative;
  }
}

@media only screen and (max-width:768px) {
  .c3 {
    position: relative;
    max-height: none;
    max-width: none;
    border-radius: 0;
    height: 100vh;
    width: 100vw;
  }
}

<body
    style=""
  >
    <div>
      <div
        class="c0"
        style="min-height: 100vh;"
      />
    </div>
    <div>
      <div
        aria-hidden="false"
      >
        <div
          class="c1"
          tabindex="-1"
        >
          <div
            class="c2"
          />
          <div
            class="c3"
            data-g-portal-id="0"
            style="overflow-y: auto;"
          >
            <a
              aria-hidden="true"
              class="c4"
              tabindex="-1"
            />
            <div>
              <div
                class="c5"
              >
                <h1
                  class="c6"
                  size="1"
                >
                  openWallet.ledger.selectWallets
                </h1>
                <div
                  class="c7"
                >
                  <div
                    class="c8"
                  >
                    Derivation path
                  </div>
                  <div
                    class="c9"
                  >
                    <button
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      aria-label="Open Drop; Selected: legacy"
                      class="c10 c11"
                      disabled=""
                      id="DerivationPathSelect"
                      type="button"
                    >
                      <div
                        class="c12"
                      >
                        <div
                          class="c13"
                        >
                          <div
                            class="c14"
                          >
                            <input
                              autocomplete="off"
                              class="c15 c16"
                              id="DerivationPathSelect__input"
                              readonly=""
                              tabindex="-1"
                              type="text"
                              value="Legacy (backward compatible)"
                            />
                          </div>
                        </div>
                        <div
                          class="c17"
                          style="min-width: auto;"
                        >
                          <svg
                            aria-label="FormDown"
                            class="c18"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="m18 9-6 6-6-6"
                              fill="none"
                              stroke="#000"
                              stroke-width="2"
                            />
                          </svg>
                        </div>
                      </div>
                    </button>
                  </div>
                </div>
                <span
                  class="c19"
                >
                  <a
                    class="c20"
                    href="https://docs.oasis.dev/general/manage-tokens/faq#ledger-derivation-paths"
                    rel="noopener"
                    target="_blank"
                  >
                    Learn more
                  </a>
                </span>
                <div
                  class="c21"
                >
                  <button
                    class="c22"
                    disabled=""
                    style="border-radius: 4px;"
                    type="button"
                  >
                    openWallet.ledger.cancel
                  </button>
                  <div
                    class="c23"
                  />
                  <button
                    class="c24"
                    data-testid="ledger-open-accounts"
                    disabled=""
                    style="border-radius: 4px;"
                    type="button"
                  >
                    openWallet.ledger.openWallets
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="StyledGrommet-sc-19lkkz7-0 cIHPIM"
      style="min-height: 100vh;"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
